src/environmentd/src/http.rs

```diff
@@ -187,19 +188,134 @@ pub struct InternalHttpConfig {
     pub tracing_handle: TracingHandle,
     pub adapter_client_rx: oneshot::Receiver<mz_adapter::Client>,
     pub active_connection_count: Arc<Mutex<ConnectionCounter>>,
+    pub promote_leader: mpsc::Sender<()>,
+    pub ready_to_promote: mpsc::Receiver<()>,
 }
 
 pub struct InternalHttpServer {
     router: Router,
 }
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct LeaderStatusResponse {
+    pub status: LeaderStatus,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LeaderStatus {
+    /// The current leader returns this. It shouldn't need to know that another instance is attempting to update it.
+    IsLeader,
+    /// The new pod should return this status until it is ready to become the leader, or it has determined that it cannot proceed.
+    Initializing,
+    /// Once we receive this status, we can tell it to become the leader and migrate the EIP.
+    ReadyToPromote,
+}
+
+#[derive(Debug)]
+pub enum LeaderState {
+    IsLeader,
+    Initializing {
+        promote_leader: mpsc::Sender<()>,
+        ready_to_promote: mpsc::Receiver<()>,
+    },
+    ReadyToPromote {
+        promote_leader: mpsc::Sender<()>,
+    },
+}
+
+fn state_to_status(state: &LeaderState) -> LeaderStatus {
+    match state {
+        LeaderState::IsLeader => LeaderStatus::IsLeader,
+        LeaderState::Initializing { .. } => LeaderStatus::Initializing,
+        LeaderState::ReadyToPromote { .. } => LeaderStatus::ReadyToPromote,
+    }
+}
+
+pub async fn handle_leader_status(
+    State(state): State<Arc<Mutex<LeaderState>>>,
+) -> impl IntoResponse {
+    let mut leader_state = state.lock().expect("lock poisoned");
+    match &mut *leader_state {
+        LeaderState::IsLeader => (),
+        LeaderState::Initializing {
+            promote_leader,
+            ready_to_promote,
+        } => {
+            match ready_to_promote.try_recv() {
+                Ok(_) => {
+                    *leader_state = LeaderState::ReadyToPromote {
+                        promote_leader: promote_leader.clone(),
+                    };
+                }
```

Comment:
ahhh, I see why you made this an mpsc @mjibson, this is tricky. You could also make `promote_leader: Option<oneshot::Sender>` so we can move it out of one enum variant and into the other. I would _slightly_ prefer this, but it's not a big deal

Reply:
Also in favor of the Option if it works! Using mpsc made me sad.

Reply:
Pushed a diff for Option<oneshot>. there are two `expect`s but not a high price to pay. None of these options is great

Reply:
Also found https://github.com/Sgeo/take_mut/blob/master/src/lib.rs and https://docs.rs/replace_with/latest/replace_with/fn.replace_with_or_abort.html, neither of which are already dependencies that would make this a little nicer (maybe?)

Reply:
Those crates both use unsafe. I don't think the risk is worth it.